Run an dummy rpc._all_gather in init_rpc to avoid shutdown timeout

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#59801 Run an dummy rpc._all_gather in init_rpc to avoid shutdown timeout**

Fixes https://github.com/pytorch/pytorch/issues/59795.

The RPC calls in shutdown no longer able to finish within 5s if
there is no other RPCs before `rpc.shutdown()` in that process,
because agent initialization can take longer than 5s. We don't
have this problem previously, because TensorPipe's backend
registry used to use RPC to communicate CUDA devices in `init_rpc`.
However, after #58753, `init_rpc` uses ProcessGroup to communicate
devices, and hence the channels/transport could be uninitialized
after `init_rpc`.

Differential Revision: [D29039238](https://our.internmc.facebook.com/intern/diff/D29039238)

mrshenli committed:
Increase shutdown timeout

1st attempt to address https://github.com/pytorch/pytorch/issues/59795.

The RPC calls in shutdown no longer able to finish within 5s. It could
be relevant to the additional checks and verifications we added for
RRef wrapped values. After increasing the timeout, the error disappeared
in my local dev env. I would like to test if this is sufficient for
CI test too.

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29039238).

cbalioglu commented:
Do we have an open GitHub issue for a long-term fix? 30 seconds seems fairly excessive for a shutdown operation.

rohan-varma reviewed (approved):
Stamp to unblock because it looks like tests pass, but would be good have a follow up issue to track a more robust fix as @cbalioglu mentioned.

mrshenli committed:
Update on "Increase shutdown timeout"


1st attempt to address https://github.com/pytorch/pytorch/issues/59795.

The RPC calls in shutdown no longer able to finish within 5s. It could
be relevant to the additional checks and verifications we added for
RRef wrapped values. After increasing the timeout, the error disappeared
in my local dev env. I would like to test if this is sufficient for
CI test too.

Differential Revision: [D29039238](https://our.internmc.facebook.com/intern/diff/D29039238)

[ghstack-poisoned]

mrshenli committed:
Update on "Run an dummy rpc._all_gather in init_rpc to avoid shutdown timeout"


Fixes https://github.com/pytorch/pytorch/issues/59795.

The RPC calls in shutdown no longer able to finish within 5s if
there is no other RPCs before `rpc.shutdown()` in that process,
because agent initialization can take longer than 5s. We don't
have this problem previously, because TensorPipe's backend
registry used to use RPC to communicate CUDA devices in `init_rpc`.
However, after #58753, `init_rpc` uses ProcessGroup to communicate
devices, and hence the channels/transport could be uninitialized
after `init_rpc`.

Differential Revision: [D29039238](https://our.internmc.facebook.com/intern/diff/D29039238)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29039238).

mrshenli committed:
Update on "Run an dummy rpc._all_gather in init_rpc to avoid shutdown timeout"


Fixes https://github.com/pytorch/pytorch/issues/59795.

The RPC calls in shutdown no longer able to finish within 5s if
there is no other RPCs before `rpc.shutdown()` in that process,
because agent initialization can take longer than 5s. We don't
have this problem previously, because TensorPipe's backend
registry used to use RPC to communicate CUDA devices in `init_rpc`.
However, after #58753, `init_rpc` uses ProcessGroup to communicate
devices, and hence the channels/transport could be uninitialized
after `init_rpc`.

Differential Revision: [D29039238](https://our.internmc.facebook.com/intern/diff/D29039238)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29039238).

mrshenli committed:
Update on "Run an dummy rpc._all_gather in init_rpc to avoid shutdown timeout"


Fixes https://github.com/pytorch/pytorch/issues/59795.

The RPC calls in shutdown no longer able to finish within 5s if
there is no other RPCs before `rpc.shutdown()` in that process,
because agent initialization can take longer than 5s. We don't
have this problem previously, because TensorPipe's backend
registry used to use RPC to communicate CUDA devices in `init_rpc`.
However, after #58753, `init_rpc` uses ProcessGroup to communicate
devices, and hence the channels/transport could be uninitialized
after `init_rpc`.

Differential Revision: [D29039238](https://our.internmc.facebook.com/intern/diff/D29039238)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29039238).

mrshenli committed:
Update on "Run an dummy rpc._all_gather in init_rpc to avoid shutdown timeout"


Fixes https://github.com/pytorch/pytorch/issues/59795.

The RPC calls in shutdown no longer able to finish within 5s if
there is no other RPCs before `rpc.shutdown()` in that process,
because agent initialization can take longer than 5s. We don't
have this problem previously, because TensorPipe's backend
registry used to use RPC to communicate CUDA devices in `init_rpc`.
However, after #58753, `init_rpc` uses ProcessGroup to communicate
devices, and hence the channels/transport could be uninitialized
after `init_rpc`.

Differential Revision: [D29039238](https://our.internmc.facebook.com/intern/diff/D29039238)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29039238).

mrshenli committed:
Update on "Run an dummy rpc._all_gather in init_rpc to avoid shutdown timeout"


Fixes https://github.com/pytorch/pytorch/issues/59795.

The RPC calls in shutdown no longer able to finish within 5s if
there is no other RPCs before `rpc.shutdown()` in that process,
because agent initialization can take longer than 5s. We don't
have this problem previously, because TensorPipe's backend
registry used to use RPC to communicate CUDA devices in `init_rpc`.
However, after #58753, `init_rpc` uses ProcessGroup to communicate
devices, and hence the channels/transport could be uninitialized
after `init_rpc`.

Differential Revision: [D29039238](https://our.internmc.facebook.com/intern/diff/D29039238)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29039238).

lw commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -353,6 +353,15 @@ def _tensorpipe_init_backend_handler(store, name, rank, world_size, rpc_backend_
+    api._all_gather(None, timeout=rpc_constants.DEFAULT_RPC_TIMEOUT_SEC)
+    # Need a barrier here to make sure no peers leave before the rank0 finishes
+    # _all_gather
+    group.barrier().wait()
```
I'm wondering: would this also work if we only did the barrier, _without_ the `_all_gather` call? That way we would prevent any worker from advancing until all workers have finished initializing their agents.

mrshenli replied:
No, after removing that `all_gather`, I still hit the following error. And this is because, without running RPC here, some ranks' `shutdown()` will be the first place to launch RPC. And the first RPC on one rank is very slow due to initialization delay, and will lead to the timeout error.  

```
Failed to respond to 'Shutdown Proceed' in time, got error RPCErr:1:RPC ran for more than set timeout (5000 ms) and will now be marked with an error
Failed to respond to 'Shutdown Proceed' in time, got error RPCErr:1:RPC ran for more than set timeout (5000 ms) and will now be marked with an error
```
